ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -339,31 +311,72 @@ static VirtualizationRootHandle FindRootAtVnode_Locked(vnode_t vnode, uint32_t v
     return RootHandle_None;
 }
 
-// Returns negative value if it failed, or inserted index on success
-static VirtualizationRootHandle InsertVirtualizationRoot_Locked(PrjFSProviderUserClient* userClient, pid_t clientPID, vnode_t vnode, uint32_t vid, FsidInode persistentIds, const char* path)
+static VirtualizationRootHandle FindOrInsertVirtualizationRoot_LockedMayUnlock(vnode_t virtualizationRootVNode, uint32_t rootVid, FsidInode persistentIds, const char* path)
 {
-    VirtualizationRootHandle rootIndex = FindUnusedIndexOrGrow_Locked();
-    
-    if (RootHandle_None != rootIndex)
+    VirtualizationRootHandle rootIndex;
+    do
     {
-        assert(rootIndex < s_maxVirtualizationRoots);
-        assert(!s_virtualizationRoots[rootIndex].inUse);
-        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
+        rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, persistentIds);
+        if (rootIndex >= 0)
+        {
+            return rootIndex;
+        }
         
-        root->providerUserClient = userClient;
-        root->providerPid = clientPID;
-        root->inUse = true;
-
-        root->rootVNode = vnode;
-        root->rootVNodeVid = vid;
-        KextLog_File(vnode, "InsertVirtualizationRoot_Locked: virtualization root inserted at index %d: (path: \"%s\", fsid: 0x%x:%x, inode: 0x%llx) directory vnode %p:%u, user client PID %d, IOUC %p.",
-            rootIndex, path, persistentIds.fsid.val[0], persistentIds.fsid.val[1], persistentIds.inode, KextLog_Unslide(vnode), vid, clientPID, KextLog_Unslide(userClient));
+        rootIndex = FindUnusedIndex_Locked();
+        if (rootIndex < 0)
+        {
+            // No space, resize array
+            uint16_t newLength = MIN(s_maxVirtualizationRoots * 2u, INT16_MAX + 1u);
```

Comment:
I don't think we'll ever get this high, but for my own understanding do you remember why `INT16_MAX + 1` selected?

Reply:
`VirtualizationRootHandle` is typedef'd to an `int16_t`,  so we can have roots `0`…`INT16_MAX`.